src/samplers/mh.jl

```diff
@@ -0,0 +1,243 @@
+      runmodel(model, vi, nothing)
+    end
+    old_θ = copy(vi[spl])
+    old_logp = getlogp(vi)
+
+    dprintln(2, "Propose new parameters from proposals...")
+    propose(model, spl, vi)
+
+    dprintln(2, "computing accept rate α...")
+    α = getlogp(vi) - old_logp + spl.info[:proposal_ratio]
+
+    dprintln(2, "decide wether to accept...")
+    if log(rand()) < α && !spl.info[:violating_support]  # accepted
+      push!(spl.info[:accept_his], true)
+    else                      # rejected
+      push!(spl.info[:accept_his], false)
+      vi[spl] = old_θ         # reset Θ
+      setlogp!(vi, old_logp)  # reset logp
+    end
+
+    vi
+  end
+end
+
+function sample(model::Function, alg::MH;
+                save_state=false,         # flag for state saving
+                resume_from=nothing,      # chain to continue
+                reuse_spl_n=0,            # flag for spl re-using
+                )
+
+  spl = reuse_spl_n > 0 ?
+        resume_from.info[:spl] :
+        Sampler(alg)
+  alg_str = "MH"
+
+  # Initialization
+  time_total = zero(Float64)
+  n = reuse_spl_n > 0 ?
+      reuse_spl_n :
+      alg.n_iters
+  samples = Array{Sample}(n)
+  weight = 1 / n
+  for i = 1:n
+    samples[i] = Sample(weight, Dict{Symbol, Any}())
+  end
+
+  vi = resume_from == nothing ?
+            Base.invokelatest(model, VarInfo(), nothing) :
+            resume_from.info[:vi]
+
+  if spl.alg.gid == 0
+    runmodel(model, vi, spl)
+  end
+
+  # MH steps
+  if PROGRESS spl.info[:progress] = ProgressMeter.Progress(n, 1, "[$alg_str] Sampling...", 0) end
+  for i = 1:n
+    dprintln(2, "$alg_str stepping...")
+
+    time_elapsed = @elapsed vi = step(model, spl, vi, i == 1)
+    time_total += time_elapsed
+
+    if spl.info[:accept_his][end]     # accepted => store the new predcits
+      samples[i].value = Sample(vi, spl).value
+    else                              # rejected => store the previous predcits
+      samples[i] = samples[i - 1]
+    end
+    samples[i].value[:elapsed] = time_elapsed
+
+    if PROGRESS ProgressMeter.next!(spl.info[:progress]) end
+  end
+
+  println("[$alg_str] Finished with")
+  println("  Running time        = $time_total;")
+  accept_rate = sum(spl.info[:accept_his]) / n  # calculate the accept rate
+  println("  Accept rate         = $accept_rate;")
+
+  if resume_from != nothing   # concat samples
+    unshift!(samples, resume_from.value2...)
+  end
+  c = Chain(0, samples)       # wrap the result by Chain
+  if save_state               # save state
+    save!(c, spl, model, vi)
+  end
+
+  c
+end
+
+function rand_truncated(dist, lowerbound, upperbound)
+    notvalid = true
+    x = 0.0
+    while (notvalid)
+        x = rand(dist)
+        notvalid = ((x < lowerbound) | (x > upperbound))
+    end
+    return x
+end
+
+assume(spl::Sampler{MH}, dist::Distribution, vn::VarName, vi::VarInfo) = begin
+    if isempty(spl.alg.space) || vn.sym in spl.alg.space
+      vi.index += 1
+      old_val = getval(vi, vn)[1]
```

Comment:
`getval(vi, vn)` is returning an array of size 1 for a univariate random variable, therefore I have to do `[1]` to get the real value.
But that's ugly and I'm pretty sure that would fail for the multivariate case.
@yebai would you know what should I do ?

Reply:
@xukai92 is there any reasons why `getval(vi, vn)` has to return an array?

Reply:
`getval(vi, vn)` calls `view(vi.vals[end], getrange(vi, vn))` which returns `1-element SubArray`.
You can try `view([1,2], 1:1)` for instance.

Reply:
Should I use `vi[vn]` instead ?

Reply:
yes, I think calling `vi[vn]` is the prefered way in Turing. `getval` is supposed to be for internal use for `VarInfo` related functions only.